JOSS paper

RubelMozumder opened the pull request:
Preparation for JOSS paper submission:
1. This inlcudes the assets with raw files and image pdf/SVG/png
2. Has a build script to build the pdf and review it.
3. Other two files paper.md and paper.bib mainly to submit to JOSS

RubelMozumder committed:
First commit.

lukaspie commented on paper/paper.md:
```diff
@@ -0,0 +1,64 @@
+    - NeXus
+    - NXDL
+authors:
+  - name: Adrian M. Price-Whelan
```
Add correct authors here, see https://github.com/FAIRmat-NFDI/pynxtools/pull/658/files for most ORCIDS etc.

mkuehbach replied:
exactly, in addition use the 
 equal-contrib for the following authors:
 * Rubel (many implementation details)
 
 we should negotiate as to how we order these three authors
 * Andrea (handover from Markus further functionalities)
 * Lukas (many bugfixes and improved tests)
 * Markus (original implementation from scratch plus bugfixes later), 
 
 all remaining authors do not have the equal contributions

RubelMozumder replied:
It could be something like:
1. Rubel, 2. Lukas, 3. Markus 4. Andrea

mkuehbach replied:
@RubelMozumder please make your proposal straight via a commit, suggestions were specific, so cycle should be implement, commit, review, implement, commit, review, ..., resolve, merge

RubelMozumder replied:
All the area-B members are equal contributors.

mkuehbach commented on paper/paper.md:
```diff
@@ -0,0 +1,65 @@
+---
+title: 'nyaml: A Format Converter for NeXus Data Model.'
+tags:
```
add:
YAML
XML

RubelMozumder replied:
added!

mkuehbach commented on paper/paper.md:
```diff
@@ -0,0 +1,65 @@
+---
+title: 'nyaml: A Format Converter for NeXus Data Model.'
```
"A" not to be capitalized

add "the" after "for" i.e. "for the NeXus"

mkuehbach commented on paper/paper.md:
```diff
@@ -0,0 +1,65 @@
+# Statement of need
+# NYAML to NXDL
+# NXDL to NYAML
+# Evaluation from NAIC
```
Evaluation from the NIAC

mkuehbach replied:
We shall though not have these acronyms in the section header unless we are sure to have these introduced in preceeding sections, overall though spelling these acronyms out is better is my opinion

RubelMozumder replied:
Abbreviation of the acronyms is included.

mkuehbach commented on paper/paper.md:
```diff
@@ -0,0 +1,65 @@
+- `[@author1:2001; @author2:2001]` -> "(Author1 et al., 2001; Author2 et al., 2002)"
+
+# Figures
+Note! The follwoing is example figure
```
see comment for citation section above

mkuehbach commented on paper/paper.md:
```diff
@@ -0,0 +1,65 @@
+Figure sizes can be customized by adding an optional second parameter:
+![Caption for example figure.](figure.png){ width=20% }
+
+# Acknowledgements
```
add # Author contributions

RubelMozumder replied:
Mainly focused on the funding agency (FAIRmat), as Nomad and other FAIRmat papers are acknowledge.

mkuehbach reviewed (changes requested):
Do not merge before submission to the journal!

This branch is the proposal and draft branch that all feature branches for the paper should use as a target!

lukaspie committed:
JOSS paper: summary and statement of needs (#77)

Co-authored-by: RubelMozumder <32923026+RubelMozumder@users.noreply.github.com>

lukaspie committed:
add a CI/CD workflow for building the paper draft (#72)

sanbrock committed:
joss NIAC (#75)

* rebase

* Update paper.md

* Update paper.md

---------

Co-authored-by: Rubel <rubel.mozumder@outlook.com>

RubelMozumder committed:
Bring all together.

RubelMozumder committed:
draft

RubelMozumder committed:
reorder the authors.

update paper and figure.

Update with build.sh file.

Update.

minor update.

RubelMozumder commented on paper/paper.md:
```diff
@@ -0,0 +1,64 @@
+   index: 2
+ - name: Independent Researcher, Country
+   index: 3
+date: 13 August 2017
```
Update!

RubelMozumder commented:
Needs to be fixed
==============================================
- [x] Check all the references, e.g., reference for xml is missing.
- [x] Bring the consistent syntax e.g. replace `\verb| nyaml |` to `__nyaml__` 
- [x] Rephrase the `summary` and `statement of need`
- [x] Fix broken author in citation

RubelMozumder committed:
Fix missing reference, check consistency for keyword and fix brocken author in reference.

RubelMozumder committed:
Little update of sammary and Statement of need. (#79)

* Little update of sammary and Statement of need.

* Update paper/paper.md

Co-authored-by: Lukas Pielsticker <50139597+lukaspie@users.noreply.github.com>

---------

Co-authored-by: Lukas Pielsticker <50139597+lukaspie@users.noreply.github.com>

lukaspie commented on paper/paper.md:
```diff
@@ -0,0 +1,122 @@
+    surname: Mozumder
+    orcid: https://orcid.org/0009-0007-5926-6646
+    affiliation: 1
+    equal-contrib: true
```
We cannot have 5 people with equal contribution! That is saying that five people are the lead author.

I think for this particular paper we have agreed that Rubel is the sole main author. So please remove equal-contrib everywhere

RubelMozumder replied:
Yeah, initially, I put all the contributors in a simple list. But interestingly, in NOMAD, all the area (D) members are equal contributors. It is nice, no! That's the reason I put all the members from area B as equal contributors. Basically, almost all of us joined in the discussion for defining grammar for NYAML.

[Nomad-joss-paper!](https://joss.theoj.org/papers/10.21105/joss.05388). But if all of you (@mkuehbach and @sherjeelshabih) want, I can remove the equal-contributor term.

lukaspie commented on paper/paper.md:
```diff
@@ -0,0 +1,122 @@
+    equal-contrib: true
+
+affiliations:
+  - name: Physics Department and CSMB, Humboldt-Universität zu Berlin, Berlin, Germany
```
Should have full addresses, see pynxtools draft

lukaspie commented on paper/paper.md:
```diff
@@ -0,0 +1,122 @@
+
+# nyaml Converter
+
+The __nyaml__ tool is a Python package, containing the several modules, that provides a command line interface for converting NeXus application definitions or base classes from YAML format (file with a __.yaml__ extension) into the XML format (file with a __.nxdl.xml__ extension) and vice-versa. To write a NeXus application defintion or a base class in YAML format, the __nyaml__ package introduce ...
```
The first sentence is just a repetition of what is said in line 79

RubelMozumder replied:
Refactored!

lukaspie commented on paper/paper.md:
```diff
@@ -0,0 +1,122 @@
+
+The __nyaml__ tool is a Python package, containing the several modules, that provides a command line interface for converting NeXus application definitions or base classes from YAML format (file with a __.yaml__ extension) into the XML format (file with a __.nxdl.xml__ extension) and vice-versa. To write a NeXus application defintion or a base class in YAML format, the __nyaml__ package introduce ...
+
+ The __nyaml__ package is designed to be used as a command line tool, but it can also be utilised as a Python module for programmatic use. The package in published in PyPI and therefore can be installed using python package manager __pip__. The converter command is invoked by the __nyaml2nxdl__ registered in the __nyaml.cli__ module [@Click:2025]. In that module, the function __launch_tool__ deci ...
```
```suggestion
 The __nyaml__ package is designed to be used as a command line tool, but it also contains Python module for programmatic use. The package in published in PyPI and therefore can be installed using python package managers (e.g.  __pip__. ` The converter command is invoked by the __nyaml2nxdl__ registered in the __nyaml.cli__ module [@Click:2025]. In that module, the function __launch_tool__ decides upon the input file type which conversion (either from YAML to XML or XML to YAML) will be invoked and execute the corresponding pipline of the data workflow (\autoref{fig:nyaml_workflow}).
```

RubelMozumder replied:
Addressed!

lukaspie reviewed (changes requested):
I had a look at the draft for the main part of the paper. I think this still needs a lot of work:
- Instead of describing the internal structure of the package (i.e., where different functions are located(, we should talk about what it does and what the logic is
- There are still lots of typos. These are very tedious to review. I strongly suggest to use a spellchecker and/or a grammar tool.
- I don't think we need to have two separate sections for YAML->XML and XML->YAML, we can probably do this together.

RubelMozumder committed:
Fix the equal contributors list. (#80)

RubelMozumder committed:
Fixed typos.

RubelMozumder commented:
> I had a look at the draft for the main part of the paper. I think this still needs a lot of work:
> 
> * Instead of describing the internal structure of the package (i.e., where different functions are located(, we should talk about what it does and what the logic is


> * There are still lots of typos. These are very tedious to review. I strongly suggest to use a spellchecker and/or a grammar tool.

Typos are fixed and spell checker is implemented in ci/cd.

> * I don't think we need to have two separate sections for YAML->XML and XML->YAML, we can probably do this together.

Yeah. Subsections may help follow the paper without going into the details.

lukaspie commented on project-words.txt:
```diff
@@ -0,0 +1,64 @@
+Koennecke
+NIAC
+NIAC
+modernising
```
We should not use British English, let's change to "modernizing" int the text and remove it from this list

mkuehbach replied:
American English is to be used throughout FAIRmat work

lukaspie commented on project-words.txt:
```diff
@@ -0,0 +1,64 @@
+Nxdl
+lxml
+lxml
+NIAC
```
There's NIAC here at least 5 times, make the list unique

RubelMozumder committed:
Refactored the converter text based on the workflow

RubelMozumder commented:
> I had a look at the draft for the main part of the paper. I think this still needs a lot of work:
> 
> * Instead of describing the internal structure of the package (i.e., where different functions are located(, we should talk about what it does and what the logic is

The text is refactored, saying less about the internal code structure rather focused how logics are implemented. @lukaspie please take a look.

> * There are still lots of typos. These are very tedious to review. I strongly suggest to use a spellchecker and/or a grammar tool.
> * I don't think we need to have two separate sections for YAML->XML and XML->YAML, we can probably do this together.

RubelMozumder committed:
Update review converter concepts.

Resolve PR comments.

PR comment resolved

PR comment resolved

PR comment resolved

Resolve PR comments.

mkuehbach commented on .cspell/cspell.config.yaml:
Why not use this one ? https://github.com/FAIRmat-NFDI/nexus_definitions/blob/fairmat/cspell.json

RubelMozumder replied:
Both are doing the same configuration, only using `yaml` instead of `json`.

mkuehbach commented on .cspell/custom-dictionary.txt:
having a few subcategories would be good, it is good that the list is sorted in ascending order lexicographically

RubelMozumder replied:
Yeah, currently it has `paper` category of documents, later we will have others like `code`, docs etc and they would be alphabetically ordered!

mkuehbach commented on .cspell/custom-dictionary.txt:
```diff
@@ -0,0 +1,41 @@
+Großen
+Heiko
+José
+Koennecke
```
Könnecke

RubelMozumder replied:
I do not think there is anything to comment on in this file. The words are just referring words in the mkdocs file. If you find the consistency of words in the `md` files, they will be reflected here.

mkuehbach replied:
His surname is "Könnecke", therefore, as commented also in the manuscript, please write "Könnecke", hence also in the spellchecker we should write "Könnecke"

lukaspie replied:
I suspect this comes from the way that the references are written (Koennecke), so we probably need to keep it.

mkuehbach commented on .cspell/custom-dictionary.txt:
```diff
@@ -0,0 +1,41 @@
+Sandor
+Shabih
+Sherjeel
+Staudtstr
```
Staudtstraße

mkuehbach replied:
We should spell out streetnames in full, abbreviating these is a German thing

RubelMozumder replied:
cspell is loogin for word `Staudtstra` and added.

mkuehbach commented on paper/build.sh:
Where is this taken from?

lukaspie replied:
This is directly from the JOSS website, I also use this locally

mkuehbach commented on paper/paper.bib:
Some missing pieces of information for some references, see https://github.com/FAIRmat-NFDI/pynxtools/blob/joss_paper/paper/paper.bib which may help to complement some like click

mkuehbach commented on paper/paper.bib:
```diff
@@ -0,0 +1,70 @@
+}
+
+@software{lxml:2025,
+    author = {{The lxml development team}},
```
Behner et al.

RubelMozumder replied:
I also noticed the name in their website. Behner stated the `lxml`, but later it was embraced and contributed by a bigger community, including `Behner`. It is not an good idea only saying `Behner et al`.

mkuehbach replied:
If there are authors we should name them, we unfortunately do not have the space to name them all that's why we use Behner et al., mentioning anybody is not only inconsistent but also poor if we know who the people are.

mkuehbach commented on paper/paper.bib:
```diff
@@ -0,0 +1,70 @@
+    url = {https://github.com/lxml/lxml}
+}
+
+@article{Koennecke:2015,
```
Könnecke

mkuehbach replied:
also add Könnecke2006 and Klosowski1997

RubelMozumder replied:
I have not found such papers nexus mentioned in their website.

lukaspie replied:
```bibtex
@article{Klosowski1997,
title = {NeXus: A common format for the exchange of neutron and synchroton data},
journal = {Physica B: Condensed Matter},
volume = {241-243},
pages = {151-153},
year = {1997},
note = {Proceedings of the International Conference on Neutron Scattering},
issn = {0921-4526},
doi = {https://doi.org/10.1016/S0921-4526(97)00865-X},
url = {https://www.sciencedirect.com/science/article/pii/S092145269700865X},
author = {Klosowski, Przemek and K{\"{o}}nnecke, Mark and Tischler, Jonathan Z. and Osborn, Ray},
}
```
and 

```bibtex
@article{Koennecke2006,
title = {The state of the NeXus data format},
journal = {Physica B: Condensed Matter},
volume = {385-386},
pages = {1343-1345},
year = {2006},
issn = {0921-4526},
doi = {https://doi.org/10.1016/j.physb.2006.06.106},
url = {https://www.sciencedirect.com/science/article/pii/S0921452606013779},
author = {K{\"{o}}nnecke, Mark},
}
```

mkuehbach replied:
These are the ones, and also here please citationkey "KönneckeSOMETHING" cuz when we deal with German Umlaut we should do so consistently

RubelMozumder replied:
Thanks @lukaspie those are added, but I was wondering when NeXus did not rigorously mention those papers, how important to add them here.

RubelMozumder replied:
Added! just a note: In the author list it is written as `K{\"{o}}nnecke`, but the name you are talking that was a tag.

mkuehbach commented on paper/paper.bib:
```diff
@@ -0,0 +1,70 @@
+  keywords = {dsl yaml},
+  month = {9},
+  timestamp = {2011-02-07T17:06:48.000+0100},
+  title = {{YAML} Ain't Markup Language ({YAML}) (tm) Version 1.2},
```
we are not using version 1.2 so we should also add one refer for an earlier if existent

RubelMozumder replied:
Any idea, what version we are using?

lukaspie replied:
pyyaml uses v1.1

lukaspie replied:
https://pypi.org/project/PyYAML/

RubelMozumder replied:
Updated!

mkuehbach commented on paper/paper.md:
```diff
@@ -0,0 +1,126 @@
+affiliations:
+  - name: Physics Department and CSMB, Humboldt-Universität zu Berlin, Zum Großen Windkanal 2, D-12489 Berlin, Germany
+    index: 1
+  - name: Lehrstuhl für Angewandte Physik, Friedrich-Alexander-Universität Erlangen-Nürnberg, Staudtstr. 7, D-91058 Erlangen, Germany
```
Staudtstra{\ss}e

RubelMozumder replied:
This is how the addresses are given in institution's website and
```
Lehrstuhl für Angewandte Physik
Friedrich-Alexander-Universität Erlangen-Nürnberg
Staudtstr. 7 / Bau A3
91058 Erlangen
```
Also, using the same address in `pynxtools` paper.

mkuehbach replied:
Correct and also there I kindly asked Sherjeel to write the address out like I ask here.
We should not type and discuss more than it takes us time to fix it.

mkuehbach commented on paper/paper.md:
```diff
@@ -0,0 +1,126 @@
+
+# Summary
+
+The NeXus scientific data format standard [@Koennecke:2015], which was originally introduced for neutron, X-ray, and muon science, has in recent years seen a significant enhancement across diverse scientific domains such as materials science. NeXus definitions specify the hierarchical structure and semantics of valid NeXus files and are written in XML [@xml10] using the NeXus Definition Language  ...
```
Könnecke2015,Könnecke2006,Klosowski1997

mkuehbach commented on paper/paper.md:
```diff
@@ -0,0 +1,126 @@
+
+# Evaluation from NIAC
+
+The NeXus International Advisory Committee (NIAC) is the governing body responsible for overseeing the development and maintenance of the NeXus data standard. A core responsibility of the NIAC is the stewardship of the NeXus Definition Language (NXDL), the XML-based schema that defines the hierarchical structure and semantics of NeXus data files [Koennecke:2015]. As part of its mission to facilit ...
```
"the XML schema" > "the XML schemas"

RubelMozumder replied:
addresed!

mkuehbach commented on paper/paper.md:
```diff
@@ -0,0 +1,126 @@
+
+# Evaluation from NIAC
+
+The NeXus International Advisory Committee (NIAC) is the governing body responsible for overseeing the development and maintenance of the NeXus data standard. A core responsibility of the NIAC is the stewardship of the NeXus Definition Language (NXDL), the XML-based schema that defines the hierarchical structure and semantics of NeXus data files [Koennecke:2015]. As part of its mission to facilit ...
```
recently, we should replace with link to the pr where this was decided

mkuehbach replied:
"Following a successful" this sentence is a duplicate just continue with "In support of..."

mkuehbach replied:
Not sure whether we should explain the addition of the two maekfile targets that long maybe just


In support of this decision [add ref to that pr or release] two makefile targets were added emdash make nyaml and make nxdl emdash for converting to YAML and XML respectively.

mkuehbach replied:
The last sentence is a testimonial statement that we may shorten but indeed could keep on the other hand if the standardization body wouldnt do anything it is dead

lukaspie replied:
Regarding the first comment, I think this is the PR: https://github.com/nexusformat/definitions/pull/1336

lukaspie replied:
Add a link to the https://github.com/nexusformat/definitions when you talk about the official definitions

mkuehbach commented on paper/paper.md:
```diff
@@ -0,0 +1,126 @@
+The NeXus International Advisory Committee (NIAC) is the governing body responsible for overseeing the development and maintenance of the NeXus data standard. A core responsibility of the NIAC is the stewardship of the NeXus Definition Language (NXDL), the XML-based schema that defines the hierarchical structure and semantics of NeXus data files [Koennecke:2015]. As part of its mission to facilit ...
+
+# Acknowledgements
+The __nyaml__ software development is funded by the German National Research Data Infrastructure
```
we should check that this follows exactly the parameterized statement requested by the dfg

RubelMozumder replied:
This is the example: 
```Example: "This work was supported by the National Research Data Infrastructure (NFDI) consortium [NFDI consortium name] – [Reference number]" ``` used for it.

mkuehbach commented on paper/paper.md:
```diff
@@ -0,0 +1,126 @@
+5. The converter writes the YAML file and appends both the hash and the original XML content as comments at the end of the YAML file.
+
+
+By attaching the hash and the original XML content to the YAML output (__.yaml__ file), the tool enables lossless round-trip conversions — provided the YAML content remains unchanged. That is, if the YAML is not modified, converting back from YAML to XML, the original commented XML content will be written back to the XML file without any modification. However, if the YAML content is modified, the ...
```
see comment I made on the docstring branch https://github.com/FAIRmat-NFDI/nyaml/pull/33/files#r2202707647

mkuehbach replied:
"That is .." sentence remove rather we should make the previous sentence stating this clearly

mkuehbach commented on paper/paper.md:
```diff
@@ -0,0 +1,126 @@
+The growth of both the standard and the number of NeXus definition developers makes it all the more important to ensure that the development process is both user-friendly and resilient to errors. The existing representation of NeXus definitions in XML offers structural rigor through the NeXus Definition Language (NXDL) and a well-defined hierarchy for metadata and data types. However, it is verbo ...
+
+
+# nyaml Converter
```
Conversion process

mkuehbach commented on paper/paper.md:
```diff
@@ -0,0 +1,126 @@
+
+
+# nyaml Converter
+
```
First paragraph shorten, everything that was  already mentioned at this point remove

lukaspie replied:
changed

mkuehbach commented on paper/paper.md:
```diff
@@ -0,0 +1,126 @@
+
+# nyaml Converter
+
+The __nyaml__ tool is a Python package developed for converting NeXus application definitions or base classes from YAML format (files with a __.yaml__ extension) into the XML format (files with a __.nxdl.xml__ extension), and vice versa. The package is published on PyPI and therefore can be installed using Python package managers (e.g., __pip__). To write a NeXus application definition or base cl ...
```
pip pypi link to add

RubelMozumder replied:
added!

mkuehbach commented on paper/paper.md:
```diff
@@ -0,0 +1,126 @@
+The __nyaml__ tool is a Python package developed for converting NeXus application definitions or base classes from YAML format (files with a __.yaml__ extension) into the XML format (files with a __.nxdl.xml__ extension), and vice versa. The package is published on PyPI and therefore can be installed using Python package managers (e.g., __pip__). To write a NeXus application definition or base cl ...
+
+
+The tool is designed to be used as a command-line utility but can also be utilized as a Python module for programmatic usages. The converter command is invoked using __nyaml2nxdl__, which determines the conversion direction based on the input file (either from YAML to XML or XML to YAML) and delegates the task to the appropriate converter. The converter then executes the corresponding data workfl ...
```
Could start with the sentence introducing the figure, the sentence "These workflows..." is imprecise what makes it efficient? reliable? reproducible? and integrity preserving so should be removed

RubelMozumder replied:
`efficient`: SHA256 hash mechanism makes the conversion efficient, where we build neither the `yaml` object nor the `xml` tree, but produces the `xml` file.

`reliable and reproducible`: The tool can create the same `xml` file if nothing is changed in `yaml` and does not lose any comments.

`integrity`: One can easily extend the existing nexus schemas with the help of the tools that integrates the new physical concept with the existing concepts.

RubelMozumder replied:
Added the name of the workflows to clarify "These workflows..."

mkuehbach commented on paper/paper.md:
```diff
@@ -0,0 +1,126 @@
+The conversion algorithm interprets the specific keywords and syntactic rules to transcode the NXDL from YAML into XML (see [documentation](https://fairmat-nfdi.github.io/nyaml)). Leveraging the NXDL rules, the conversion process detects possible inconsistencies in the YAML content and raises errors or warnings if the rules are not properly followed.
+
+
+The __XML to YAML__ conversion also follows a well-defined data workflow (\autoref{fig:nyaml_workflow}) that converts a given input XML file into a YAML file. The workflow begins with the XML input and proceeds as follows:
```
Move the XML to yaml block first,

always one starts with xml to yaml because still xml is the preferred choice

RubelMozumder replied:
The workflow in a order, `yaml->xml` and `xml->yaml`, it is fair to keep the order consistent.
I really not not understand how the order makes a difference. we are talking about a tool that converts from yaml ->xml and xml->  yml. It does not matter who came first and who came later in description.

lukaspie replied:
Suggestion: add a sentence describing the difference use cases

- When editing existing definitions, workflow is XML ->YAML -> XML
- When making a new definition from scratch, you _can_ go YAML -> YML directly

mkuehbach reviewed (dismissed):
Commented

lukaspie commented on paper/paper.md:
```diff
@@ -0,0 +1,125 @@
+---
+# Summary
+
+The NeXus scientific data format standard [@Könnecke:2015]  [@Könnecke2006] [@Klosowski1997], which was originally introduced for neutron, X-ray, and muon science, has in recent years seen a significant enhancement across diverse scientific domains such as materials science. NeXus definitions specify the hierarchical structure and semantics of valid NeXus files and are written in XML [@xml10] usi ...
```
Have a look at how to use multiple references: https://joss.readthedocs.io/en/latest/example_paper.html

They should be separated by `;`

RubelMozumder committed:
Fix diagram and marmaid plot.

RubelMozumder committed:
minor change.

RubelMozumder committed:
Fix the custom-dictionary path.

Update paper/paper.md

Co-authored-by: Lukas Pielsticker <50139597+lukaspie@users.noreply.github.com>

Update paper/paper.md

Co-authored-by: Lukas Pielsticker <50139597+lukaspie@users.noreply.github.com>

Update paper/paper.md

Co-authored-by: Lukas Pielsticker <50139597+lukaspie@users.noreply.github.com>

Update paper/paper.md

Co-authored-by: Lukas Pielsticker <50139597+lukaspie@users.noreply.github.com>

Update paper/paper.md

Co-authored-by: Lukas Pielsticker <50139597+lukaspie@users.noreply.github.com>

Update paper/paper.md

Co-authored-by: Lukas Pielsticker <50139597+lukaspie@users.noreply.github.com>

Update paper/paper.md

Co-authored-by: Lukas Pielsticker <50139597+lukaspie@users.noreply.github.com>

lukaspie commented on paper/paper.md:
```diff
@@ -0,0 +1,99 @@
+---
+# Summary
+
+The NeXus scientific data format standard [@Könnecke:2015; @Könnecke2006; @Klosowski1997], which was originally introduced for neutron, X-ray, and muon science, has in recent years seen a significant enhancement across diverse scientific domains such as materials science. NeXus definitions, consisting of application definitions or base classes, specify the hierarchical structure and semantics of  ...
```
Maybe we can have a reference for XSD?

RubelMozumder replied:
Added!

lukaspie commented on paper/paper.md:
```diff
@@ -0,0 +1,99 @@
+
+The NeXus scientific data format standard [@Könnecke:2015; @Könnecke2006; @Klosowski1997], which was originally introduced for neutron, X-ray, and muon science, has in recent years seen a significant enhancement across diverse scientific domains such as materials science. NeXus definitions, consisting of application definitions or base classes, specify the hierarchical structure and semantics of  ...
+
+__nyaml__ is a Python-based tool with both a command-line interface and an importable API that facilitates the conversion between NXDL XML and a simplified YAML [@YAML11] representation. YAML's indentation-based syntax enhances human readability and simplifies manual editing. By providing a reliable, lossless round-trip conversion between XML and YAML, __nyaml__ enables developers to edit NeXus d ...
```
I suggest we change naming of software tools using ``, like in the [JOSS example paper](https://joss.readthedocs.io/en/latest/example_paper.html). So, for this part:
```suggestion
`nyaml` is a Python-based tool with both a command-line interface and an importable API that facilitates the conversion between NXDL XML and a simplified YAML [@YAML11] representation. YAML's indentation-based syntax enhances human readability and simplifies manual editing. By providing a reliable, lossless round-trip conversion between XML and YAML, `nyaml` enables developers to edit NeXus definitions more efficiently without sacrificing structural or semantic fidelity.
```

and same below

lukaspie commented on paper/paper.md:
```diff
@@ -0,0 +1,99 @@
+__nyaml__ is a Python-based tool with both a command-line interface and an importable API that facilitates the conversion between NXDL XML and a simplified YAML [@YAML11] representation. YAML's indentation-based syntax enhances human readability and simplifies manual editing. By providing a reliable, lossless round-trip conversion between XML and YAML, __nyaml__ enables developers to edit NeXus d ...
+
+# Statement of need
+The growth of both the standard and the number of NeXus definition developers makes it all the more important to ensure that the development process is both user-friendly and resilient to errors. The existing representation of NeXus definitions in XML offers structural rigor through the NeXus Definition Language (NXDL) and a well-defined hierarchy for metadata and data types. However, it is verbo ...
```
```suggestion
The growth of both the standard and the number of NeXus definition developers makes it all the more important to ensure that the development process is both user-friendly and resilient to errors. The existing representation of NeXus definitions in XML offers structural rigor through the NeXus Definition Language (NXDL) and a well-defined hierarchy for metadata and data types. However, it is verbose and can be difficult to edit by hand. Writing and maintaining NXDL files often involves dealing with deeply nested elements and strict syntax, which can be error-prone and time-consuming, especially for users who are not familiar with XML development. `nyaml` addresses these challenges by enabling the development of NeXus definitions in a simpler, YAML-based format while preserving the full structure, semantics, and developer comments of the original XML.
```

lukaspie commented on paper/paper.md:
```diff
@@ -0,0 +1,99 @@
+
+# Conversion process
+
+The __nyaml__ tool is a Python package developed for converting NeXus definitions from YAML format (files with a __.yaml__ extension) into the XML format (files with a __.nxdl.xml__ extension), and vice versa. The package is published on PyPI and therefore can be installed using Python package managers (e.g., __pip__ [@pip2025]). To write or modify NeXus definitions in YAML format, modification o ...
```
```suggestion
The `nyaml` tool is a Python package developed for converting NeXus definitions from YAML format (files with a `.yaml` extension) into the XML format (files with a `.nxdl.xml` extension), and vice versa. The package is published on PyPI and therefore can be installed using Python package managers (e.g., `pip` [@pip2025]). The `nyaml` package introduces a set of keywords and syntactic rules (see [documentation](https://fairmat-nfdi.github.io/nyaml)) that are specific to NXDL (NeXus Definition Language) in YAML. Existing definitions can be modified using the following workflow: __XML $\rightarrow$ YAML $\rightarrow$ modification of YAML $\rightarrow$ XML__. New definitions can be designed in YAML and converted to XML directly 
```

lukaspie commented on paper/paper.md:
```diff
@@ -0,0 +1,99 @@
+The __nyaml__ tool is a Python package developed for converting NeXus definitions from YAML format (files with a __.yaml__ extension) into the XML format (files with a __.nxdl.xml__ extension), and vice versa. The package is published on PyPI and therefore can be installed using Python package managers (e.g., __pip__ [@pip2025]). To write or modify NeXus definitions in YAML format, modification o ...
+
+
+The tool is designed to be used as a command-line utility but can also be utilized as a Python module for programmatic usages. The converter command is invoked using __nyaml2nxdl__, which determines the conversion direction based on the input file (either from YAML to XML or XML to YAML) and delegates the task to the appropriate converter. The converter then executes the corresponding data workfl ...
```
```suggestion
The tool is designed to be used as a command-line utility, but can also be utilized as a Python module for programmatic usages. The converter command is invoked using `nyaml2nxdl`, which determines the conversion direction based on the input file (either from YAML to XML or XML to YAML) and delegates the task to the appropriate converter. The converter then executes the corresponding data workflow pipeline (see \autoref{fig:nyaml_workflow}). These workflows, YAML to XML and XML to YAML. are designed to ensure that the conversion process is efficient, reliable, reproducible, and preserves the integrity of the original NeXus data structure and semantics.
```

lukaspie commented on paper/paper.md:
```diff
@@ -0,0 +1,99 @@
+
+![Converter workflow for nyaml tool, XML to YAML and vice-versa \label{fig:nyaml_workflow}](assets/workflow-1.pdf){ width=75% }
+
+Conversion from __YAML to XML__ follows specific workflow steps (depicted in \autoref{fig:nyaml_workflow}). Given an input YAML file (see \autoref{fig:nyaml_workflow}), the __nyaml2nxdl__ converter reads the __.yaml__ file, collects and tracks the locations of comments, and then constructs a nested Python hash-mapped object representing the YAML content using __PyYAML__ [@PyYAML:2024]. In the fin ...
```
```suggestion
Conversion from __YAML to XML__ follows specific workflow steps (depicted in \autoref{fig:nyaml_workflow}). Given an input YAML file (see \autoref{fig:nyaml_workflow}), the `nyaml2nxdl` converter reads the `.yaml` file, collects and tracks the locations of comments, and then constructs a nested Python hash-mapped object representing the YAML content using `PyYAML` [@PyYAML:2024]. In the final stage, in accordance with the NXDL grammar and syntax specific to the XML format, the converter generates an XML file that includes the collected comments. In the __YAML to XML__ conversion, the algorithm interprets the keywords and syntactic rules specific to the YAML format (see [documentation](https://fairmat-nfdi.github.io/nyaml)) to transcode the NXDL content from YAML into XML. Leveraging the NXDL rules, the conversion process detects possible inconsistencies in the YAML content and raises errors or warnings if the rules are not properly followed.
```

lukaspie commented on paper/paper.md:
```diff
@@ -0,0 +1,99 @@
+
+Conversion from __YAML to XML__ follows specific workflow steps (depicted in \autoref{fig:nyaml_workflow}). Given an input YAML file (see \autoref{fig:nyaml_workflow}), the __nyaml2nxdl__ converter reads the __.yaml__ file, collects and tracks the locations of comments, and then constructs a nested Python hash-mapped object representing the YAML content using __PyYAML__ [@PyYAML:2024]. In the fin ...
+
+In the reverse direction — __XML to YAML__ conversion (see \autoref{fig:nyaml_workflow}) — the __nxdl2nyaml__ converter takes the input __.nxdl.xml__ file and transforms it into an XML tree structure using the __lxml__ library [@lxml:2025]. This XML tree is then transcoded into a YAML file that adheres to the NXDL rules and syntax specific to the YAML format. The converter then computes a SHA256  ...
```
```suggestion
In the reverse direction — __XML to YAML__ conversion (see \autoref{fig:nyaml_workflow}) — the `nxdl2nyaml` converter takes the input `.nxdl.xml` file and transforms it into an XML tree structure using the `lxml` library [@lxml:2025]. This XML tree is then transcoded into a YAML file that adheres to the NXDL rules and syntax specific to the YAML format. The converter then computes a SHA256 hash from the generated YAML content and appends the hash to the end of the YAML file, followed by the original XML content. By attaching both the hash and the original XML content to the YAML output (`.yaml` file), the tool enables lossless round-trip conversions: if the YAML is not modified, then  the original commented XML content will be restored in the output XML file without modification when converting from YAML to XML. However, if the YAML content is changed, the XML tree will be reconstructed from the modified YAML and written to the XML file, resulting in differences from the original XML content included as comments. This caching approach streamlines the XML → YAML → XML workflow and facilitates straightforward comparisons of XML files in version control systems such as Git.
```

lukaspie commented on paper/paper.md:
```diff
@@ -0,0 +1,99 @@
+
+# Evaluation from NIAC
+
+The NeXus International Advisory Committee (NIAC) is the governing body responsible for overseeing the development and maintenance of the NeXus data standard. A core responsibility of the NIAC is the stewardship of the NeXus Definition Language (NXDL), the XML-based definitions that define the hierarchical structure and semantics of NeXus data files [Könnecke:2015]. As part of its mission to faci ...
```
```suggestion
The NeXus International Advisory Committee (NIAC) is the governing body responsible for overseeing the development and maintenance of the NeXus data standard. A core responsibility of the NIAC is the stewardship of the NeXus Definition Language (NXDL), the XML-based definitions that define the hierarchical structure and semantics of NeXus data files [Könnecke:2015]. As part of its mission to facilitate the standardization of NeXus definitions in NXDL, NIAC has recently reviewed and formally accepted the `nyaml` tool. Following a successful evaluation, NIAC has approved `nyaml` and endorsed it as the recommended tool for preparing NeXus definition proposals. In support of this decision, the official NeXus definition repository [@nexusDef:2024] was updated to integrate `nyaml` into its workflow through the addition of two makefile targets: 'make nyaml' [@nyamlIntegration:2024], which converts existing definitions from the canonical nxdl.xml format into .yaml, and 'make nxdl', which detects modified or newly added .yaml files and converts them back into valid nxdl.xml format for submission and version control. This integration ensures that contributions made in .nyaml are compatible with the existing XML-based infrastructure. The adoption of `nyaml` by NIAC reflects an ongoing commitment to fostering community engagement and modernizing the technical tools underpinning the NeXus standard [@NIAC:2025].
```

lukaspie reviewed (commented):
Looks good now, some very small comments

RubelMozumder committed:
add xsd reference.

lukaspie commented on paper/paper.md:
```diff
@@ -0,0 +1,99 @@
+The NeXus International Advisory Committee (NIAC) is the governing body responsible for overseeing the development and maintenance of the NeXus data standard. A core responsibility of the NIAC is the stewardship of the NeXus Definition Language (NXDL), the XML-based definitions that define the hierarchical structure and semantics of NeXus data files [Könnecke:2015]. As part of its mission to faci ...
+
+# Acknowledgements
+The __nyaml__ software development is funded by the German National Research Data Infrastructure
```
```suggestion
The `nyaml` software is developed by FAIRmat. FAIRmat is funded by the Deutsche Forschungsgemeinschaft (DFG, German Research Foundation) – project 460197019.
```

lukaspie commented on paper/paper.md:
```diff
@@ -0,0 +1,99 @@
+---
+# Summary
+
+The NeXus scientific data format standard [@Könnecke:2015; @Könnecke2006; @Klosowski1997], which was originally introduced for neutron, X-ray, and muon science, has in recent years seen a significant enhancement across diverse scientific domains such as materials science. NeXus definitions, consisting of application definitions or base classes, specify the hierarchical structure and semantics of  ...
```
add "xmlschema" to custom-dictionary

RubelMozumder committed:
Update funding statement and minor fix.

lukaspie commented:
@domna Hey Florian, we did not have your private email address, so I am trying to reach you here. We have finalized paper drafts for the nyaml tool (here) and for pynxtools (see https://github.com/FAIRmat-NFDI/pynxtools/pull/651). We have you as a coauthor on both (since you significantly contributed to both packages) and would love to here your opinion. Feel free to reach out via email if that works better.

domna commented:
Hey @lukaspie, thank you for reaching out. Nice that the papers are getting in shape 🎉. For me github is fine (if there is something where I should have a look just address me directly), but I'll also share my private e-mail address with you.

RubelMozumder committed:
Fix ref for Florian.

RubelMozumder committed:
Address comment from Nathan.

RubelMozumder commented:
Commit d193828 addresses the comment from Nathan
```
This implies that the YAML format retains the XML tags and fields. Where is this format specified? Are XML comments special tags or keywords here?
```

RubelMozumder committed:
Feedback from Pepe.

RubelMozumder committed:
Claudia's comment.

RubelMozumder committed:
Extend space in yaml workflow.

RubelMozumder committed:
Apply cludia's suggestion.

mkuehbach commented on paper/paper.md:
```diff
@@ -0,0 +1,98 @@
+
+# Statement of need
+
+As the NeXus standard and its community of definition developers continue to grow, it is increasingly important to ensure that the development process remains both user-friendly and resilient to errors. The existing representation of NeXus definitions in XML offers structural rigor through the NeXus Definition Language (NXDL) and a well-defined hierarchy for metadata and data types. However, it i ...
```
"As a use case, the tool has been used to integrate NeXus definitions into the large-scale research data management platform NOMAD [@NOMAD2023], with bidirectional YAML–XML conversions enabling modifications and extensions to make the NeXus schemata compatibility with the NOMAD data model and other physics communities."

reformulate like so:
"As a successful stress test, the tool was used for developing extensions of the NeXus definitions for different microscopy and spectroscopy methods to integrate these into the research data management system NOMAD [@NOMAD2023]."

"was used" or "has been used" I don't mind, use what fits best with the rest of the paragraph

mkuehbach replied:
Given that we already use references in the summary we could also move this statement on NOMAD in between "YAML representation" and "YAML integration" into the summary/abstract. I think talking about an application in an statement of needs is not a perfect fit while making that use case an own small paragraph is too little a text to justify that paragraph.

RubelMozumder replied:
Text is updated as you suggested.

mkuehbach commented on paper/paper.bib:
```diff
@@ -0,0 +1,127 @@
+    url          = {https://github.com/pallets/click}
+}
+@software{lxml:2025,
+    author = {{Behner et al}},
```
et al.

RubelMozumder replied:
Addressed!

mkuehbach commented on paper/paper.bib:
```diff
@@ -0,0 +1,127 @@
+author = {K{\"{o}}nnecke, Mark},
+}
+@misc{nyamlIntegration:2024,
+  author = {{Dobener et al.}},
```
Dobener, Florian et al.

RubelMozumder replied:
Addressed!

mkuehbach commented on paper/paper.bib:
```diff
@@ -0,0 +1,127 @@
+volume = {8},
+number = {90},
+pages = {5388},
+author = {{Scheidgen et al.}},
```
Scheidgen, M. et al., better also cite here all authors, submission process will tell us if we need to abbreviate

RubelMozumder replied:
Addressed!

mkuehbach commented on paper/assets/Nyaml_workflow.svg:
@RubelMozumder looks much better now, good job.

I recommend to move the itle "Choose the Right converter" a little bit down to have as best as possible the same spacing between the top boundary of the box and its title, respectively in each box.

mkuehbach replied:
Also recommend to not repeat Converter in the boxes talking about the "Right Converter" it is implicit that these are converters.

mkuehbach replied:
Personally I would prefer if the boxes "Comment Collector" and "XML Object" have their upper edge aligned to the same height, likewise the pair "Python dictionary object" and "YAML file intermediate"

mkuehbach replied:
The figure caption should make bold the statement XML to YAML to be consistent with the writing in the paper body.

RubelMozumder replied:
We never used the `XML to YAML` bold in the paper. But I make the full caption bold.

RubelMozumder committed:
update workflow svg image.

RubelMozumder committed:
Markus comments.

RubelMozumder committed:
Update pdf

RubelMozumder committed:
update image .

lukaspie commented on cspell.config.yaml:
```diff
@@ -1,10 +0,0 @@
----
```
Why is this file removed?

RubelMozumder replied:
added!

lukaspie commented on paper/assets/mermaidPlot.md:
```diff
@@ -0,0 +1,33 @@
+    subgraph Start
+      id1["Input File (YAML or XML)"]
+    end
+    subgraph Correct File Converter
```
I suggest to change to "Choice of File Converter"

RubelMozumder replied:
Updated!

RubelMozumder commented on paper/assets/workflow-1.pdf:
Delete this file.

RubelMozumder replied:
Done!

lukaspie commented on paper/assets/mermaidPlot.md:
````diff
@@ -0,0 +1,33 @@
+  ```mermaid
````
Remove this file if it is not used anymore.

RubelMozumder replied:
Updated!

RubelMozumder committed:
Update files.

RubelMozumder committed:
update paper.pdf.

mkuehbach commented:
**This branch needs to considered as frozen for adding new features as a submission is in process!**

RubelMozumder committed:
Fix doi and orcid id

RubelMozumder reviewed (commented):
(Outdated) Delete the preprint files on discussions.

RubelMozumder committed:
Research Impact Statement, Software design, and AI usage disclosure.

RubelMozumder committed:
Includes: Research Impact Statement, Software design, and AI usage disclosure.

RubelMozumder committed:
Add State of the field.

RubelMozumder committed:
Fix typos and add missing word in cspell dictionary.

RubelMozumder committed:
Rebase-1

RubelMozumder committed:
Fix Atom prob microscopy

RubelMozumder committed:
Include published paper.

RubelMozumder committed:
Remove duplication of cspell file and use the common one json config file

RubelMozumder committed:
Removed paper draft github action.

lukaspie reviewed (approved):
LGTM! Thank you and congrats for seeing the paper submission through 😃
